Added vision.org and initial proposal

hadjian opened the pull request.

barnstee commented on design/0002-vscode-extension-wot-cli.org:
```diff
@@ -0,0 +1,39 @@
+#+TITLE: VSCode Extension and wot CLI
```
Not sure I believe in the VSCode extension, as the persona developing the WoT TD is not a developer but an Ot operator, in my world. Therefore, I really like the fact that the edi[TD]or is hosted for you and you can integrate it into any toolchain onboarding workflow easily...

hadjian replied:
Good point for a brown-field on-boarding scenario. Probably it makes sense to write down scenarios that people have in their mind.

We oftentimes write a couple of TMs and some algorithm that works on on the generated TDs. That would be easier in an integrated environment, with TMs, TD generation and runtime.

Also there are device integrators that write TMs in an back-office for 3rd party devices. They will have a repository of TMs and probably property templates that they will want to access.

Maybe we can get some input from others on the roles of people who will write TMs/TDs.

egekorkan replied:
So regarding integration into a workflow:

We need to understand what people expect from such TD editors and tools. From my point of view, EdiTDor is now a Web UI so it cannot be integrated into a workflow per se but its specific features can be or it can be used (not integrated) as a tool in a workflow (like Postman). Integratabiltiy is what [TD Playground](https://github.com/thingweb/thingweb-playground) does where different packages are offered as a Node.js dependency, [CLI](https://github.com/thingweb/thingweb-playground/tree/master/packages/cli) or integrated in a [Web UI](http://plugfest.thingweb.io/playground/). 

Anyways I think a separate discussion/issue is needed for this point :)

rhofer replied:
@hadjian I think it would be worth to invite for a short call to layout what we already have and do (and open source), as:
- TD Playground mentioned by @egekorkan above, where @edbingo did a full rework of the cli td validator under the lead of @mkovatsc + all the other stuff on TD Playground @egekorkan and others are doing anyways (also with @sebastiankb towards w3c standardization)
- the ediTDor of course
- then beginning of this year @ektrah open sourced a type script based ontology viewer, running in browser with [rdf-toolkit](https://github.com/ektrah/rdf-toolkit)
- ... maybe more

All this has the potential to be used together / integrate / run side-by-side ... whatever we can imagine and makes sense

To make a statement here for the VSC extension per se, I'm not to deep in the daily usage (and pain) of TM/TD handling. I'd need here the input from @fennibay , @mkovatsc and further colleagues doing this already.

But anyhow, my two cents on this: VSC extension's main benefit is to do developers life easier, where you code and have TM/TD side-by-side and would enjoy a great support handling them by an extension. And if this applies, and let's assume an extension is created, don't be surprised for receiving more features to be delivered. (as e.g. TM/TD object based comparison of two TM/TD's (diffing))

hadjian replied:
Great pointers. I will invite for a one to one meeting and put a summary here. And your mention of this side-by-side workflow is what I am interested in.

gintephSICKAG replied:
I agree with @barnstee. 
Having the ediTDor integrated into our Thing Model repository is most beneficial for us. In this way, we have an easy to use single source of truth for Thing Models.

fennibay replied:
VSC Extension would be helpful for development cases, I agree.

What we'd like even more is a reusable version of editdor that we can integrate in our engineering software, e.g. as a web component. We can also talk to Ozgur Gulderen about this, who is working on such an engineering software.

relu91 replied:
I join the comment train :). Internally we use the [td-code](https://marketplace.visualstudio.com/items?itemName=arces-wot.td-code) and we usually edit what I would call [Partial TDs](https://www.w3.org/TR/wot-architecture11/#dfn-partial-td) alongside the code (see an old experiment [here](https://github.com/vaimee/WoTSimFarm/tree/master/backend) - it is also a good example of how you can use node-wot as a runtime+cli). I would say that this workflow is now a little bit outdated cause it predates the introduction of Thing Models, but in general, I expect that to stay and I have a couple of inputs about what I would like to see in a full-fledge VSCode extension:
- Intellisense for Json-ld --> meeting understanding terms imported in contexts (vscode currently supports only minimal features meaning it uses the JSON extension on json-ld files see https://github.com/microsoft/vscode/issues/95013 and https://github.com/microsoft/vscode/issues/56334)
- Intellisense for ThingModels --> meaning I can get autocomplete and validation from the linked TMs
- I usually don't like visual tooling when coding, but for some people, it might be helpful to have a rendered preview of the TM/PartialTD with the linked TMs (it could accessible by the usual ctrl+shift+v shortcut)

hadjian replied:
Perfect. It seems very clear that the ediTDor approach is working, so we should start there and improve. Also, there seem to be quite some usable components available that I didn't know of. @egekorkan is giving me a tour today. I will adjust the proposal and get back to all of you. Good thing we are doing this open source. The feedback is pure gold ;-)

fennibay commented on design/0002-vscode-extension-wot-cli.org:
```diff
@@ -0,0 +1,39 @@
+- Create a VSCode extension that syntactically validates TMs/TDs
+- Context extensions of JSON-LD may be used to map to JSON Schemas
+- Provide wot-ui to create HTML pages for TMs/TDs
+- Introduce a wot-cli
+- wot-cli can validate TMs/TDs against JSON Schemas
+- wot-cli can generate stubs for a protocol driver to communicate with it
+- wot-cli can be used to communicate with an implementation of those stubs
+- wot-cli can observe properties and events and record them in timeseries
+- wot-cli can generate appropriate CSV protocol bindings
+- wot-cli can add protocol bindings to TMs to exchange field level bindings with CSV ones
```
I would align wot-cli with [@thingwebplayground/cli](https://github.com/thingweb/thingweb-playground/tree/master/packages/cli). As that is already doing some of the functionality listed here, @edbingo is also working on extensions. It might make sense to keep editdor as an editor, and leave cli tasks to thingwebplayground.

hadjian replied:
Thanks for your feedback @fennibay. I got an introduction to the existing thingweb project by @egekorkan and I agree with your points. I'll think about how to document these insights.

hadjian replied:
@fennibay we'll go into detail about what task is fulfilled by which component in the scenarios we will document. I agree that https://github.com/thingweb/thingweb-playground/tree/master/packages/cli should be used. The ediTDor uses this project already for validation and this is how the tools should work together. I think there is room for improvement in the way it is used.

egekorkan commented on media/wot-workflow.drawio.svg:
I would try to separate starting from commit to catalog since now it seems that the one who commits to the catalog and the one using a Thing from the catalog have to be the same person.

hadjian replied:
Ok, I also noticed this separation when talking to @fennibay. He called the first blocks "design time" and the second set of blocks (run time), if I remember correctly. Will change that and remove the arrow? Would that reflect your intent?

egekorkan replied:
Removing the arrow would make sense. We can also separate them based on the roles such as committer and user/consumer.

hadjian replied:
@egekorkan Updated the diagram. Can you confirm that this is what you had in mind? I wouldn't go into roles yet, as it will assume too much. The workflows are quite different as of now.

egekorkan replied:
It is better to have them separated like they are now. 

For future discussions:

At least in my point of view, there are two run times: run time of the Thing and run time of the application code. Until a Thing is fully commissioned, there is no run time of the application code which is also part of WoT. Thus, this workflow would be called "Thing Workflow" rather than "WoT Workflow". We can have another workflow discussion later on for workflows on writing the application code and combined they can be named the WoT workflow. 


Note: I think a better word should be found asap for "Run Time" since it can be quickly confused with "Runtime", which is overloaded already.

hadjian replied:
- Deployment Time? As in https://arc42.org/overview
- Bootstrapping/Provisioning/Commissioning? https://www.w3.org/TR/wot-architecture11/#thing-lifecycle, but this would exclude writing application code
- Operation Time? As an enclosing term for everything that makes use of an already existing TM
- Usage Time? Also enclosing as the point above.

egekorkan replied:
> Bootstrapping/Provisioning/Commissioning? https://www.w3.org/TR/wot-architecture11/#thing-lifecycle, but this would exclude writing application code

I think this fits the most and writing application code has a complete separate cycle and different toolset. It would be good the hear opinions on this 👀

hadjian replied:
Hmmm, don't know if we need to be that exact right now. My opinion is to keep things simple as this is not an rigorous standards text, but just a big picture. We can change it, if we see the need for it. What do you think about "design" and "usage" to subsume the two big blocks. I don't think it will do harm for the first baby steps we'll take.

egekorkan replied:
Yup I think that's fine and we can have further discussion on this later on.

hadjian replied:
https://github.com/web-of-things-open-source/proposal/blob/0001-wot-oss-vision/media/wot-workflow.drawio.svg

egekorkan commented on media/wot-components.drawio.svg:
Why not node-wot and [WADE](https://github.com/tum-esi/wade) for testing with one device step?

egekorkan replied:
For the generating TDs step, one can use the [thing-model-helpers](https://github.com/eclipse-thingweb/node-wot/blob/master/packages/td-tools/src/thing-model-helpers.ts) from td-tools of node-wot (soon to be a separate repo)

hadjian replied:
These are possibilities, but we shouldn't discuss it right now. There will be separate proposals for these components, where we can discuss each component. The elements marked red are just the result of brainstorming and not set in stone.

Before we discuss components, we also need to agree on a first workflow, which only involves a subset of components and then go into proposals.

hadjian replied:
@egekorkan Updated the diagram

hadjian commented:
I realized that I shouldn't mix the vision and my suggestion where to start, so I removed the proposal and left only the vision:

- renamed and rephrased the issue
- moved the vision.org document to design/0002-wot-oss-vision.org
- renamed this pull request
- components diagram and proposal removed

Could you please comment on the vision text. If it is fine, we can use the workflow as a frame for scenarios. Appreciate timely responses, as I want to sketch out the first scenario.

egekorkan commented on vision.org:
Some overall comments:
- I was not aware that we were talking about the future of the ediTDor so far so I am a bit surprised with the title. Since I am also part of Eclipse Foundation, it might be fine but contributions to Eclipse projects should be discussed within Eclipse repos (unless we are talking about a new project?).
- It would be better to not mix "passive/we" with "I"

hadjian replied:
Yes, it is rather about exanding on what the ediTDor has started. But it is mostly about new tools and projects that will work well with the ediTDor. If that means that these projects should automatically go under the Eclipse umbrella is something I like to discuss beforehand. I will open a separate issue.

hadjian replied:
Will rephrase to not mix "I" and "we"

hadjian replied:
https://github.com/web-of-things-open-source/proposal/commit/1abdb344aad1f547f36bbcf116c1466007d25a92
- Not mixing "I" and "we"

egekorkan commented on vision.org:
```diff
@@ -0,0 +1,32 @@
+
+In this open source project, we set out to change this. The vision is to bridge the gap between the two scenarios. We will focus on the developer experience and if we succeed, an IoT developer will be able to create or better just fetch device twins, model the context, write code and deploy to production.
+
+The vision is grand, but we need to start small. At Siemens we published the open source [[https://github.com/eclipse/editdor][ediTDor]] project, which helps to write and syntactically validate TMs/TDs. We will pick this up and drive it towards an awesome Digital Twin development environment.
```
Siemens did not publish the EdiTDor open source project, Eclipse Foundation did, to which people with Siemens affiliation have contributed to. Since https://projects.eclipse.org/projects/iot.editdor/who does not even show Siemens (due to Committers not choosing their affiliation), this should be reworded.

hadjian replied:
Sure, it's hard to see that for me, since I have the inside view of this project, but you are right: there is no Siemens affiliation mentioned and this is actually what we want. We want a community driven project. Will change that.

hadjian replied:
https://github.com/web-of-things-open-source/proposal/commit/1abdb344aad1f547f36bbcf116c1466007d25a92
- Replaced Siemens with Eclipse Foundation

barnstee commented on vision.org:
```diff
@@ -0,0 +1,32 @@
+
+1. Timeseries Data: data coming from sensors, oftentimes called signals, runtime data, samples or others
+
+2. Contextual Data: this type of data answers questions about types of devices, interpretation of timeseries values, topology of assets and locations, taxonomies and more. Sometimes this type of data is called the semantics.
```
... or metadata.

hadjian replied:
https://github.com/web-of-things-open-source/proposal/commit/1abdb344aad1f547f36bbcf116c1466007d25a92
- Added "metadata" as alternative to "semantics"

barnstee commented on vision.org:
```diff
@@ -0,0 +1,32 @@
+
+2. Contextual Data: this type of data answers questions about types of devices, interpretation of timeseries values, topology of assets and locations, taxonomies and more. Sometimes this type of data is called the semantics.
+
+We do not have a consistent way to work with these kinds of data. While normal software developers happily whip out VSCode, IntelliJ, emacs or vim, install a language server and make use of a myriad of libraries to get things done, IoT developers struggle to even access and understand data, let alone deliver value on top of it.
```
application developers instead of "software developers"?

hadjian replied:
https://github.com/web-of-things-open-source/proposal/commit/1abdb344aad1f547f36bbcf116c1466007d25a92
- Rephrased "software developers" to "application developers"

barnstee commented on vision.org:
```diff
@@ -0,0 +1,32 @@
+
+In this open source project, we set out to change this. The vision is to bridge the gap between the two scenarios. We will focus on the developer experience and if we succeed, an IoT developer will be able to create or better just fetch device twins, model the context, write code and deploy to production.
+
+The vision is grand, but we need to start small. At Siemens we published the open source [[https://github.com/eclipse/editdor][ediTDor]] project, which helps to write and syntactically validate TMs/TDs. We will pick this up and drive it towards an awesome Digital Twin development environment.
```
awesome? Let's say fully-featured :-)

hadjian replied:
This was me, trying to be cool ;-)

barnstee commented on vision.org:
```diff
@@ -0,0 +1,32 @@
+
+* Target Audience
+
+/IoT Developer/ is a very abstract role and for many an over-simplification. There are developers who create protocol driver code, experts who are knowledgeable about a domain, ontologists, application developers, system integrators etc.
```
... OT developers.

hadjian replied:
https://github.com/web-of-things-open-source/proposal/commit/1abdb344aad1f547f36bbcf116c1466007d25a92
- Added "OT developers" as a role

barnstee commented on vision.org:
```diff
@@ -0,0 +1,32 @@
+
+/IoT Developer/ is a very abstract role and for many an over-simplification. There are developers who create protocol driver code, experts who are knowledgeable about a domain, ontologists, application developers, system integrators etc.
+
+In a large corporation, these roles might be well organized and fulfilled by different people. However, I was more often than not also part of a team of 2-3 normal software developers that were responsible to go to a customer site, access devices, build up the infrastructure and write the application.
```
Remove "I was more ... than not also part of ... that were " and add "are"

hadjian replied:
https://github.com/web-of-things-open-source/proposal/commit/1abdb344aad1f547f36bbcf116c1466007d25a92
- Rephrased paragraph about the target scenario

hadjian commented:
I'll regard this thread as resolved and will merge the PR.
